### PR TITLE
Allow configuring the background and accent color

### PR DESCRIPTION
**Pull request by jMaddrell**
UI For changing and viewing background and accent colours.
image

### CODE REVIEW

File: Firmware/src/network/KnomiWebServer.cpp [hunk 1]
```diff
@@ -176,6 +176,44 @@ KnomiWebServer::KnomiWebServer(Config *config, WifiManager *manager) {
...
+    if (this->config != nullptr && this->config->getUiConfig() != nullptr) {
+      if (newAccentColor != nullptr) {
+        LV_LOG_INFO((String("Updating accentColor to: ") + newAccentColor).c_str());
+        this->config->getUiConfig()->setAccentColor(strtol(newAccentColor.c_str(), NULL, 16));
+      }
+      this->config->getUiConfig()->save();
+      req->send(200, "application/json", "{result: \"ok\"}");
+
+      delay(100);
+      ESP.restart();
```
Comment (jMaddrell) [inline Firmware/src/network/KnomiWebServer.cpp:211]: Without invalidating objects this was the easiest way to apply the changes, happy to look at invalidating objects instead if you'd prefer?

File: Firmware/src/config/UIConfig.cpp [hunk 2]
```diff
@@ -3,13 +3,13 @@
...
-  this->accentColor = prefs.getUInt("accentColor", 0xFF0000);
+  this->accentColor = prefs.getInt("accentColor", 0xFF0000);
   prefs.end();
 }
 
 void UIConfig::save() {
   Preferences prefs;
   prefs.begin(this->getConfigNamespace().c_str(), false);
-  prefs.putUInt("accentColor", this->getAccentColor());
+  prefs.putInt("accentColor", this->getAccentColor());
```
Comment (jMaddrell) [inline Firmware/src/config/UIConfig.cpp:13]: Given the conversations around fs handling I've left this as nvs for now. Once we have a decision I'm happy to move to JSON

**Review by DiverOfDark — CHANGES_REQUESTED** [review]

File: Firmware/src/network/KnomiWebServer.cpp [hunk 3]
```diff
@@ -176,6 +176,44 @@ KnomiWebServer::KnomiWebServer(Config *config, WifiManager *manager) {
...
+    String newAccentColor;
+    if (req->hasArg("accentColor")) {
+      newAccentColor = req->arg("accentColor");
+
+      if (newAccentColor.startsWith("#")) {
+        newAccentColor = newAccentColor.substring(1);
+      }
+    }
+
+    if (this->config != nullptr && this->config->getUiConfig() != nullptr) {
```
[the remark this answers is not in the document]
Comment (jMaddrell) [reply Firmware/src/network/KnomiWebServer.cpp:202]: I got a null response if I loaded the page too quickly, so just added this to guard against it

File: Firmware/src/config/UIConfig.cpp (hunk 2 above)
Comment (jMaddrell) [reply Firmware/src/config/UIConfig.cpp:13]: Let me try changing it back, I was using auto locally and noticed it was using int so changed the type

File: Firmware/src/network/KnomiWebServer.cpp [hunk 4]
```diff
@@ -176,6 +176,44 @@ KnomiWebServer::KnomiWebServer(Config *config, WifiManager *manager) {
     req->send(pResponse);
   });
 
+  pServer->on("/api/themeConfig", HTTP_GET, [&](AsyncWebServerRequest *req) {
```
[the remark this answers is not in the document]
Comment (jMaddrell) [reply Firmware/src/network/KnomiWebServer.cpp:179]: Will change

File: Firmware/src/network/KnomiWebServer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (DiverOfDark) [reply Firmware/src/network/KnomiWebServer.cpp:202]: OK, let's leave it as is in this PR, but for future we may need to think about API of config system. IMO it's not very practical when we have to check for nullptr before each member access.

File: Firmware/src/network/KnomiWebServer.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jMaddrell) [reply Firmware/src/network/KnomiWebServer.cpp]: Done

File: Firmware/src/config/UIConfig.cpp (hunk 2 above)
Comment (jMaddrell) [reply Firmware/src/config/UIConfig.cpp]: Changed back to uint, it must have been something silly I did when I first made the changes that caused issues.

**Comment by jMaddrell** [comment]
Added support for theming of background colour as well
image

**Review by DiverOfDark — APPROVED** [review]

**Pull request merged**